pallets/subtensor/tests/root.rs

```diff
@@ -697,3 +697,95 @@ fn test_weights_after_network_pruning() {
+            (20_989_747_000_000_000, 488_281),   // Eleventh halving event
+            (20_994_874_000_000_000, 244_140),   // Twelfth halving event
+            (20_997_437_000_000_000, 122_070),   // Thirteenth halving event
+            (20_998_719_000_000_000, 61_035),    // Fourteenth halving event
+            (20_999_360_000_000_000, 30_517),    // Fifteenth halving event
+            (20_999_680_000_000_000, 15_258),    // Sixteenth halving event
+            (20_999_840_000_000_000, 7_629),     // Seventeenth halving event
+            (20_999_920_000_000_000, 3_814),     // Eighteenth halving event
+            (20_999_960_000_000_000, 1_907),     // Nineteenth halving event
+            (20_999_980_000_000_000, 953),       // Twentieth halving event
+            (20_999_990_000_000_000, 476),       // Twenty-first halving event
+            (20_999_990_500_000_000, 476),
+            (20_999_995_000_000_000, 238), // Twenty-second halving event
+            (20_999_998_000_000_000, 119), // Twenty-third halving event
+            (20_999_999_000_000_000, 59),  // Twenty-fourth halving event
+            (21_000_000_000_000_000, 0),   // Total supply reached, emissions stop
+            (21_100_000_000_000_000, 0)    // Just for fun
+        ];
+
+        for (issuance, expected_emission) in expected_emissions.iter() {
+            SubtensorModule::set_total_issuance(*issuance);
+            step_block(1);
+
+            let current_emission = SubtensorModule::get_block_emission();
+            assert_eq!(
+                current_emission, *expected_emission,
+                "Incorrect emission {} at total issuance {}",
+                current_emission, issuance
+            );
+        }
+    });
+}
+
+#[test]
+fn test_get_emission_across_entire_issuance_range() {
+    new_test_ext().execute_with(|| {
+        let total_supply: u64 = pallet_subtensor::TotalSupply::<Test>::get();
+        let original_emission: u64 = pallet_subtensor::DefaultBlockEmission::<Test>::get();
+        let halving_issuance: u64 = total_supply / 2;
+        let mut step: usize = original_emission as usize;
+
+        for issuance in (0..=total_supply).step_by(step) {
+            SubtensorModule::set_total_issuance(issuance);
+
+            let issuance_f64 = issuance as f64;
+            let h = f64::log2(1.0 / (1.0 - issuance_f64 / (2.0 * halving_issuance as f64)));
+            let h = h.floor();
+            let emission_percentage = f64::powf(2.0, -h);
+
+            let expected_emission: u64 = if issuance < total_supply {
+                (original_emission as f64 * emission_percentage) as u64
+            } else {
+                0
+            };
+            assert_eq!(
+                SubtensorModule::get_block_emission(),
+                expected_emission,
+                "Issuance: {}",
+                issuance_f64
+            );
+            step = expected_emission as usize;
+        }
+    });
```

Comment:
I guess as long as this consistently passes this is fine but I suppose if there is ever a hardware difference w.r.t floating points vs the safe implementation this test _might_ fail on some random hardware. Probably fine tho

my preference tho would be that this uses safe math as well if possible